src/java/org/apache/cassandra/index/sai/disk/v1/kdtree/TraversingBKDReader.java

```diff
@@ -0,0 +1,335 @@
+            negativeDeltas[level + splitDim] = false;
+            in.setPosition(nodePosition);
+            readNodeData(false);
+        }
+
+        public void pop()
+        {
+            nodeID /= 2;
+            level--;
+            splitDim = splitDims[level];
+        }
+
+        public boolean isLeafNode()
+        {
+            return nodeID >= leafNodeOffset;
+        }
+
+        public boolean nodeExists()
+        {
+            return nodeID - leafNodeOffset < leafNodeOffset;
+        }
+
+        public int getNodeID()
+        {
+            return nodeID;
+        }
+
+        public byte[] getSplitPackedValue()
+        {
+            assert !isLeafNode();
+            assert splitPackedValueStack[level] != null : "level=" + level;
+            return splitPackedValueStack[level];
+        }
+
+        /** Only valid after pushLeft or pushRight, not pop! */
+        public int getSplitDim()
+        {
+            assert !isLeafNode();
+            return splitDim;
+        }
+
+        public long getLeafBlockFP()
+        {
+            assert isLeafNode() : "nodeID=" + nodeID + " is not a leaf";
+            return leafBlockFPStack[level];
+        }
+
+        public BytesRef getSplitDimValue()
+        {
+            assert !isLeafNode();
+            scratch.bytes = splitValuesStack[level];
+            scratch.offset = splitDim * bytesPerValue;
+            return scratch;
+        }
+
+        private void readNodeData(boolean isLeft)
+        {
+            leafBlockFPStack[level] = leafBlockFPStack[level - 1];
+
+            // read leaf block FP delta
+            if (!isLeft)
+                leafBlockFPStack[level] += in.readVLong();
+
+            if (isLeafNode())
+                splitDim = -1;
+            else
+            {
+                // read split dim, prefix, firstDiffByteDelta encoded as int:
+                int code = in.readVInt();
+                splitDim = 0;
+                splitDims[level] = splitDim;
+                int prefix = code % (1 + bytesPerValue);
+                int suffix = bytesPerValue - prefix;
+
+                if (splitValuesStack[level] == null)
+                {
+                    splitValuesStack[level] = new byte[packedBytesLength];
+                }
+                System.arraycopy(splitValuesStack[level - 1], 0, splitValuesStack[level], 0, packedBytesLength);
+                if (suffix > 0)
+                {
+                    int firstDiffByteDelta = code / (1 + bytesPerValue);
+                    if (negativeDeltas[level + splitDim])
+                        firstDiffByteDelta = -firstDiffByteDelta;
+                    int oldByte = splitValuesStack[level][splitDim * bytesPerValue + prefix] & 0xFF;
+                    splitValuesStack[level][splitDim * bytesPerValue + prefix] = (byte) (oldByte + firstDiffByteDelta);
+                    in.readBytes(splitValuesStack[level], splitDim * bytesPerValue + prefix + 1, suffix - 1);
```

Comment:
In the three lines above, I think `splitDim` is always zero (set on line 308?), and makes the multiplications unnecessary.